Add support Drupal 8

TomAbel opened the pull request:
Included ajax validation for API credentials
Includes a hook_help() for support/help.
Added more cURL error code checks with messages

Commit:
Updated module to support Drupal 8

Included support for API validation with ajax.
Includes a hook_help() for in place of a README.
Added more error code checks.

Commit:
Updated module help.

The help page is visible after the module is installed.
Pointless to include installation details after install.
Added configuration info to the help page.

Commit:
Comments, formatting

Remove nocaching option in form.
Remove filedoc comment (Drupal standard).
Add comments about how validation is done.
Added Smaily in some explanations for clarity.

Commit:
Add autoresponder placeholder

This allows for no autoresponder to be set.

Commit:
Add subdomain normalizer.

Included a subdomain normalizer when validating or saving in admin.
Converts https://81cur7rc.sendsmaily.net/ -> 81cur7rc

kaittodesk commented on smaily_for_drupal/smaily_for_drupal.module:
```diff
@@ -0,0 +1,28 @@
+    case 'help.page.smaily_for_drupal':
+      $output = '';
+      $output .= '<h3>' . t('About') . '</h3>';
+      $output .= '<p>' . t('The Smaily for Drupal module allows you to integrate the Smaily API for your site. A newsletter subscription block can be configured and placed where possible.<br> More information about Smaily is available at the Smaily <a href=http://help.smaily.com/en/support/home> Knowledgebase.') . '</a></a></p>';
```
Try to avoid having links inside translation phrases.

This can be written as:
```php
t('... More information about Smaily is available in the Smaily <a href=":knowledgebase_url">Knowledgebase</a>', array(':knowledgebase_url' => 'http://help.smaily.com/en/support/home'));
```

PS. You have 2 link closing tags outside of translation phrase. One should be removed and the other should be in the phrase.

Commit:
Edited formatting, removed version from info.yml.

Version is autogenerated, shouldn't be added manually.
Made some long lines a bit shorter.

kaittodesk commented on smaily_for_drupal/smaily_for_drupal.module:
```diff
@@ -0,0 +1,28 @@
+      $output .= '<h3>' . t('Uses') . '</h3>';
+      $output .= '<dl>';
+      $output .= '<dt>' . t('Configuring Smaily API credentials') . '</dt>';
+      $output .= '<dd>' . t("Fill out and validate the credentials provided to you by Smaily in the administration page.") . '</dd>';
```
Formatting: be consistent when using double or single quotes.

kaittodesk commented on smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php:
```diff
@@ -0,0 +1,273 @@
+/**
+ * Smaily module admin configuraiton form.
+ */
+class SmailyForDrupalAdminForm extends ConfigFormBase {
```
I think this can just be `AdminForm`. The `namespace` value should make sure that this does not conflict with any other module.

kaittodesk commented on smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php:
```diff
@@ -0,0 +1,273 @@
+   */
+  protected function getEditableConfigNames() {
+    return [
+      'smaily_for_drupal.adminsettings',
```
One could think that `adminsettings` refers to admin form settings, e.g. these settings are for configuring the admin form not `Smaily for Drupal` module. I would go with `smaily_for_drupal.settings`, as these are this module's settings.

kaittodesk commented on smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php:
```diff
@@ -0,0 +1,273 @@
+      '#markup' => '<div class="smaily_message"></div>',
+    ];
+
+    $form['smaily_container']['credentials']['smaily_domain'] = [
```
I would go with `['smaily_api_credentials']['domain']` (and similar) grouping. It makes more sense.

Commit:
Added readme for Drupal 8

Commit:
Remove wrong text

kaittodesk commented on smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php:
```diff
@@ -0,0 +1,277 @@
+    $form['smaily_container']['credentials']['smaily_domain'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Domain'),
+      '#default_value' => $config->get('smaily_domain', ''),
```
And I would change configuration option key from `smaily_domain` to `smaily_api_credentials.domain` (and apply the same pattern to rest of the configuration options), it is more explicit and avoids possible conflicts in the future. Hopefully the key is not too long.

kaittodesk commented on smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php:
```diff
@@ -0,0 +1,277 @@
+    // Fetch and update autoresponders every time form is built.
+    $ch = curl_init();
+    curl_setopt($ch, CURLOPT_URL,
+      'https://' . $domain . '.sendsmaily.net/api/workflows.php?trigger_type=form_submitted'
```
What happens if `smaily_domain` does not have a value?

TomAbel replied:
cURL will say: Could not resolve host: .sendsmaily.net

kaittodesk replied:
It would be better to not make any request if any of the API credentials are missing. It is a great way to increasing efficiency.

TomAbel replied:
The validation would stop any half-filled api credentials from being saved, even with Ajax.
Is it a just-in-case scenario? I could only see this happening when someone manually edits the mySQL table

kaittodesk replied:
OK. Maybe I don't understand how and when `buildForm` is called. Doesn't it get called when the form is rendered in the admin panel? If it does, wouldn't it still call `fetchAutoresponders` even if no settings have been saved before?

TomAbel replied:
I added an !empty($domain) check (now for all credentials).
It would still call `fetchAutoresponders` but the `if` check wouldn't start a cURL session.

kaittodesk commented on smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php:
```diff
@@ -0,0 +1,277 @@
+
+    $form['smaily_container']['submit'] = [
+      '#type' => 'submit',
+      '#value' => $this->t('Save'),
```
I think it is better if the button has information about what the user is saving. Please rename it to "Save settings".

kaittodesk commented on smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php:
```diff
@@ -0,0 +1,277 @@
+      '#title' => $this->t('Password'),
+      '#default_value' => $config->get('smaily_password', ''),
+      '#required' => TRUE,
+
```
Formatting: unnecessary whitespace.

kaittodesk commented on smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php:
```diff
@@ -0,0 +1,277 @@
+    curl_setopt($ch, CURLOPT_RETURNTRANSFER, TRUE);
+    curl_setopt($ch, CURLOPT_HTTPAUTH, CURLAUTH_BASIC);
+    curl_setopt($ch, CURLOPT_USERPWD, "$username:$password");
+    $autoresponders = json_decode(curl_exec($ch), TRUE);
```
What happens if `curl_exec` returns `FALSE`?

TomAbel replied:
I tried `json_decode(FALSE, TRUE)` it returned NULL
`if (!empty($autoresponders))` will be skipped

TomAbel replied:
Is it bad not to handle a FALSE message and rely on json_decode returning NULL

kaittodesk commented on smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php:
```diff
@@ -0,0 +1,277 @@
+   * {@inheritdoc}
+   */
+  public function validateForm(array &$form, FormStateInterface $form_state) {
+    // Validation handled by #required and in ajax function.
```
Is this method required here even if it does nothing?

TomAbel replied:
Not required by ConfigFormBase, no.

kaittodesk commented on smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php:
```diff
@@ -0,0 +1,277 @@
+    $password = trim($form_state->getValue('smaily_password'));
+    $subdomain = trim($form_state->getValue('smaily_domain'));
+    // Drupal's validation is only done with submit.
+    if (empty($username) || empty($password) || empty($subdomain)) {
```
Isn't this already done by basic form validation? Or doesn't `ConfigFormBase` handle required form fields with AJAX requests?

TomAbel replied:
It does not, no. AJAX requests need special validation

kaittodesk commented on smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php:
```diff
@@ -0,0 +1,277 @@
+    // Normalize subdomain.
+    // First, try to parse as full URL.
+    // If that fails, try to parse as subdomain.sendsmaily.net, and, then clean up subdomain and pass as is.
+    if (filter_var($subdomain, FILTER_VALIDATE_URL)) {
```
Please extract this functionality into a separate method that takes the unformatted subdomain as its value and returns formatted subdomain.

kaittodesk commented on smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php:
```diff
@@ -0,0 +1,277 @@
+    }
+    $subdomain = preg_replace('/[^a-zA-Z0-9]+/', '', $subdomain);
+
+    $ch = curl_init();
```
Please extract automations fetching into separate method.

kaittodesk commented on smaily_for_drupal/src/Form/SmailyForDrupalForm.php:
```diff
@@ -0,0 +1,126 @@
+/**
+ * Smaily newsletter signup form.
+ */
+class SmailyForDrupalForm extends FormBase {
```
I think this class can be renamed to `SubscribeForm` (or similar). Again conflicts shouldn't be an issue due to the namespacing.

ghost commented:
1. I currently added the Drupal 8 module in folder smaily_for_drupal and added a GitHub readme for Drupal 8. I think we should have the same folder versioning logic like in OpenCart.
2. Do you know if and how I could somehow cut out the cURL portion into a controller or class to not repeat it 3 times in the admin form, or is it OK? Smaily is moving to captcha eventually but still would like to know.

Ready for review.

kaittodesk commented on smaily_for_drupal/src/Form/SmailyForDrupalForm.php:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    $ch = curl_init();
+    curl_setopt($ch, CURLOPT_URL, 'https://' . $domain . '.sendsmaily.net/api/opt-in/');
```
Use API endpoint `/api/autoresponder.php`. Please take a look at the [API documentation](http://help.smaily.com/en/support/solutions/articles/16000021289-autoresponder).

TomAbel replied:
What's the difference of this and `/api/opt-in/`?

kaittodesk replied:
There is none. But why then configure Smaily API credentials if you are not going to use them? :)

TomAbel replied:
I reworked it for autoresponder.php. Autoresponder.php asks for some authorization while /opt-in/ doesn't, yes? And I already went through the trouble of getting API credentials from config, so might as well cURL to autoresponder.php
Is the current solution OK or did I misunderstand you?

kaittodesk replied:
Yes, that is mighty fine. Thanks!

`/api/autoresponder.php` uses authentication and it is more meant for background requests.  `/api/opt-in/` is more meant for embedded opt-in forms that directly submit form's contents to Smaily. We are trying to make this separation at the moment with all of our integrations.

kaittodesk commented on smaily_for_drupal/src/Form/SmailyForDrupalAdminForm.php:
```diff
@@ -0,0 +1,277 @@
+    return $ajax_response;
+
+  }
+
```
Is this whitespace a Drupal coding standard?

TomAbel replied:
Line 276 yes.

kaittodesk replied:
OK.

kaittodesk commented on README.md:
```diff
@@ -0,0 +1,25 @@
+# Smaily Drupal module
```
Use just "Smaily for Drupal".

kaittodesk commented on README.md:
```diff
@@ -0,0 +1,25 @@
+# Smaily Drupal module
+
+Smaily email marketing and automation extension plugin for Drupal.
```
"extension plugin" mean the same thing, but they do not mean anything in Drupal's context. Drupal uses "module" instead.

kaittodesk commented on README.md:
```diff
@@ -0,0 +1,25 @@
+
+Smaily email marketing and automation extension plugin for Drupal.
+Includes a simple sign-up form to add new subscribers to Smaily subscribers list.
+## Features
```
Add a newline before headings for better readability.

kaittodesk commented on README.md:
```diff
@@ -0,0 +1,25 @@
+- Select autoresponder to send automated emails.
+
+## Requirements
+- Drupal 8+
```
Do not add plus sign after Drupal 8, because that would mean that this module works with future Drupal 9 and later.

kaittodesk commented on README.md:
```diff
@@ -0,0 +1,25 @@
+
+- Place smaily_for_drupal into core/modules/ folder.
+- Install module under Extend.
+- Validate and save your credentials.
```
Please extend this a little - Validate and save your Smaily API credentials.

Commit:
Pull Request Response #1 @TODO: Readme

Commit:
Revert "Pull Request Response #1 @TODO: Readme"

This reverts commit 03a5507faf9dfeeffce4c6d81f8eea40bf41bf24.
Messed up trying to commit a backup of these files to my fork.

Commit:
Changes mentioned in Pull Request. More explicitly

Refactored link inside a t() in .module file.
Changed credentials config key schema to smaily_api_credentials.foo
Removed validateForm functions as they are not required.
Extracted automations fetching & subdomain normalization into functions.
Fixed validating & then saving resulting in an illegal choice.
Changed API endpoint /api/opt-in/ -> /api/autoresponder.php
Formatting: removed extra whitespace, consistency.
Added !empty(subdomain) check before running cURL.

ghost commented:
Question. What should be this module's parent when you are on admin/config?(Development/Media/Etc)
It's currently sitting in Web Services with RSS publishing (Drupal core).

kaittodesk commented:
> What should be this module's parent when you are on admin/config?(Development/Media/Etc)
> It's currently sitting in Web Services with RSS publishing (Drupal core).

I think it is fine. Our competition uses the same section.

kaittodesk commented:
> Do you know if and how I could somehow cut out the cURL portion into a controller or class to not repeat it 3 times in the admin form, or is it OK? Smaily is moving to captcha eventually but still would like to know.

If you are feeling that the code becomes heavy (too much code for a single function or a file) or logic starts to repeat, then it is always a good idea to start refactoring it, and in the process you might need to create separate classes/functions/methods or even need to split the code into multiple files.

With the current cURL implementation, you could build a separate model for that - a single purpose class that takes in arguments necessary to make the API request and handles the response. Something similar to the Request class you built for Smaily for Wordpress 2.x plugin.

Commit:
Check credentials before fetching autoresponders

kaittodesk commented on smaily_for_drupal/smaily_for_drupal.module:
```diff
@@ -0,0 +1,29 @@
+    case 'help.page.smaily_for_drupal':
+      $output = '';
+      $output .= '<h3>' . t('About') . '</h3>';
+      $output .= '<p>' . t('The Smaily for Drupal module allows you to integrate the Smaily API for your site. A newsletter subscription block can be configured and placed where possible.');
```
You are starting a paragraph here, but I do not see a closing tag anywhere.

TomAbel replied:
Good eye. Is there any reason for this kind of formatting Drupal uses, or is it the only way to do it?

kaittodesk replied:
I am not really sure. Drupal has a decent (although complicated) templating logic. Maybe `hook_help` is triggered in such an early stage of bootstraping that usage of template is not possible. If you'd like you can take a deep dive into how `hook_help` works, but for now I think it will do.

Commit:
Missing </p>
